analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionShouldNotBeThrownFromUnexpectedMethods.cs

```diff
@@ -85,61 +83,83 @@ protected override void Initialize(SonarAnalysisContext context)
-                    DefaultAllowedExceptions),
+                c => CheckForIssue<ConversionOperatorDeclarationSyntax>(c, cods => cods.ImplicitOrExplicitKeyword.IsKind(SyntaxKind.ImplicitKeyword), DefaultAllowedExceptions),
                 SyntaxKind.ConversionOperatorDeclaration);
         }
 
-        private void CheckForIssue<TSyntax>(SyntaxNodeAnalysisContext analysisContext,
-            Func<TSyntax, bool> isTrackedSyntax, ImmutableArray<KnownType> allowedThrowTypes)
+        private static void CheckForIssue<TSyntax>(SyntaxNodeAnalysisContext analysisContext, Func<TSyntax, bool> isTrackedSyntax, ImmutableArray<KnownType> allowedThrowTypes)
             where TSyntax : SyntaxNode
         {
             var syntax = (TSyntax)analysisContext.Node;
-            if (!isTrackedSyntax(syntax))
+            if (isTrackedSyntax(syntax))
             {
-                return;
+                ReportOnInvalidThrowStatement(analysisContext, syntax, allowedThrowTypes);
             }
+        }
+
+        private static bool IsTrackedOperator(OperatorDeclarationSyntax declaration) =>
+            TrackedOperators.Contains(declaration.OperatorToken.Kind());
+
+        private static bool IsTrackedMethod(MethodDeclarationSyntax declaration, SemanticModel semanticModel) =>
+            HasTrackedMethodOrAttributeName(declaration)
+            && semanticModel.GetDeclaredSymbol(declaration) is { } methodSymbol
+            && HasTrackedMethodOrAttributeType(methodSymbol);
 
-            ReportOnInvalidThrowStatement(analysisContext, syntax, allowedThrowTypes);
+        private static bool HasTrackedMethodOrAttributeName(MethodDeclarationSyntax declaration)
+        {
+            var name = declaration.Identifier.ValueText;
+            return name == "Equals"
+                || name == "GetHashCode"
+                || name == "ToString"
+                || name == "Dispose"
+                || name == "Equals"
+                || declaration.AttributeLists.SelectMany(list => list.Attributes).Any(x => x.ArgumentList == null && x.Name.ToStringContains("ModuleInitializer"));
         }
 
-        private void ReportOnInvalidThrowStatement(SyntaxNodeAnalysisContext analysisContext,
+        private static bool HasTrackedMethodOrAttributeType(IMethodSymbol methodSymbol) =>
+            methodSymbol.IsObjectEquals()
+            || methodSymbol.IsObjectGetHashCode()
+            || methodSymbol.IsObjectToString()
+            || methodSymbol.IsIDisposableDispose()
+            || methodSymbol.IsIEquatableEquals()
+            || IsModuleInitializer(methodSymbol);
+
+        private static bool IsModuleInitializer(IMethodSymbol methodSymbol) =>
+            methodSymbol.AnyAttributeDerivesFrom(KnownType.System_Runtime_CompilerServices_ModuleInitializerAttribute);
+
+        private static void ReportOnInvalidThrowStatement(SyntaxNodeAnalysisContext analysisContext,
```

Comment:
The method name can be confusing. The method checks both throw statements and expressions.